internal/pkg/levee/propagation/propagation.go

```diff
@@ -0,0 +1,350 @@
+	// Send has no referrers, it is only an Instruction, not a Value.
+	case *ssa.Send:
+		s.dfs(t.Chan.(ssa.Node), maxInstrReached, lastBlockVisited, false)
+
+	// These nodes' operands should not be visited, because they can only receive
+	// taint from their operands, not propagate taint to them.
+	case *ssa.BinOp, *ssa.ChangeInterface, *ssa.ChangeType, *ssa.Convert, *ssa.Extract, *ssa.MakeChan, *ssa.MakeMap, *ssa.MakeSlice, *ssa.Phi, *ssa.Range:
+		s.visitReferrers(n, maxInstrReached, lastBlockVisited)
+
+	// These nodes don't have operands; they are Values, not Instructions.
+	case *ssa.Const, *ssa.FreeVar, *ssa.Global, *ssa.Lookup, *ssa.Parameter:
+		s.visitReferrers(n, maxInstrReached, lastBlockVisited)
+
+	// These nodes don't have referrers; they are Instructions, not Values.
+	case *ssa.Go:
+		s.visitOperands(n, maxInstrReached, lastBlockVisited)
+
+	// These nodes are both Instructions and Values, and currently have no special restrictions.
+	case *ssa.Field, *ssa.MakeInterface, *ssa.Select, *ssa.Slice, *ssa.TypeAssert, *ssa.UnOp:
+		s.visitReferrers(n, maxInstrReached, lastBlockVisited)
+		s.visitOperands(n, maxInstrReached, lastBlockVisited)
+
+	// These nodes cannot propagate taint.
+	case *ssa.Builtin, *ssa.DebugRef, *ssa.Defer, *ssa.Function, *ssa.If, *ssa.Jump, *ssa.MakeClosure, *ssa.Next, *ssa.Panic, *ssa.Return, *ssa.RunDefers:
+
+	default:
+		fmt.Printf("unexpected node received: %T %v; please report this issue\n", n, n)
+	}
+}
+
+func (s *DFSRecord) visitReferrers(n ssa.Node, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
+	if n.Referrers() == nil {
+		return
+	}
+	for _, r := range *n.Referrers() {
+		s.dfs(r.(ssa.Node), maxInstrReached, lastBlockVisited, true)
+	}
+}
+
+func (s *DFSRecord) visitOperands(n ssa.Node, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
+	for _, o := range n.Operands(nil) {
+		if *o == nil {
+			continue
+		}
+		s.dfs((*o).(ssa.Node), maxInstrReached, lastBlockVisited, false)
+	}
+}
+
+func (s *DFSRecord) canReach(start *ssa.BasicBlock, dest *ssa.BasicBlock) bool {
+	if start.Dominates(dest) {
+		return true
+	}
+
+	stack := stack([]*ssa.BasicBlock{start})
+	seen := map[*ssa.BasicBlock]bool{start: true}
+	for len(stack) > 0 {
+		current := stack.pop()
+		if current == dest {
+			return true
+		}
+		for _, s := range current.Succs {
+			if seen[s] {
+				continue
+			}
+			seen[s] = true
+			stack.push(s)
+		}
+	}
+	return false
+}
+
+// String implements Stringer interface.
+func (s *DFSRecord) String() string {
+	var b strings.Builder
+	for _, n := range s.compress() {
+		b.WriteString(fmt.Sprintf("%v ", n))
+	}
+
+	return b.String()
+}
+
+// compress removes the elements from the graph that are not required by the
+// taint-propagation analysis. Concretely, only propagators, sanitizers and
+// sinks should constitute the output. Since, we already know what the source
+// is, it is also removed.
+func (s *DFSRecord) compress() []ssa.Node {
```

Comment:
Out of scope: This feels like it should be removed in the near future. I believe this was originally added to support the `String` method above, which itself was used in tests?

Reply:
I think it's in scope to prune this, since it's being moved anyway.